<a href="https://colab.research.google.com/github/Thippawan72/Thippawan-Supho/blob/main/%E0%B8%AA%E0%B8%B3%E0%B9%80%E0%B8%99%E0%B8%B2%E0%B8%82%E0%B8%AD%E0%B8%87_Frequent_Patterns_(Association_Rules).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import os

In [ ]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/MyDrive/BSC-dpdm23_flie'

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
data = pd.read_csv(os.path.join(path,'reduced_marketbasket.csv'),encoding='latin-1')

In [ ]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

In [ ]:
data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

        InvoiceDate  UnitPrice  CustomerID         Country  
0  12/01/2010 08:26       2.55     17850.0  United Kingdom  
1  12/01/2010 08:26       3.39     17850.0  United Kingdom  
2  12/01/2010 08:26       2.75     17850.0  United Kingdom  
3  12/01/2010 08:26       3.39     17850.0  United Kingdom  
4  12/01/2010 08:26       3.39     17850.0  United Kingdom

In [ ]:
type(data['Description'][0])

str

#Quit1 มีกี่ประเทศ

In [ ]:
len(set(data['Country']))

38

#เเปลง data ใก้อยู่ในรูปเเบบ Transaction data ด้วย ```groupby.apply(list) ```

In [ ]:
#data.groupby('InvoiceNo')['StockCode'].apply(list)

In [ ]:
transactions = data.groupby('InvoiceNo')['StockCode'].apply(list)

## Association Rules

In [ ]:
!pip install apyori

In [ ]:
from apyori import apriori

In [ ]:
!pip install apyori
from apyori import apriori
transactions = data.groupby('InvoiceNo')['StockCode'].apply(list)
frequentItemsets = list(apriori(transactions,min_support = 0.065))
frequentItemsets

[RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06586872586872587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'47566'}), confidence=0.06586872586872587, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.08243243243243244, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.08243243243243244, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08671814671814672, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08671814671814672, lift=1.0)])]

In [ ]:
len(frequentItemsets)

4

In [ ]:
frequentItemsets[0]

RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)])

In [ ]:
frequentItemsets[0][0]

frozenset({'22423'})

In [ ]:
dir(frequentItemsets[0][0])

['__and__',
 '__class__',
 '__class_getitem__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__ne__',
 '__new__',
 '__or__',
 '__rand__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__ror__',
 '__rsub__',
 '__rxor__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__xor__',
 'copy',
 'difference',
 'intersection',
 'isdisjoint',
 'issubset',
 'issuperset',
 'symmetric_difference',
 'union']

In [ ]:
frequentItemsets[0][0].__len__()#หาขนาดของ itemset

1

In [ ]:
min_sup = 0.03

frequentItemsets = list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22386', '85099B'}), support=0.03216216216216216, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])
yes
RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03027027027027027, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])


RelationRecord(items=frozenset({'22386', '85099B'}),
support=0.03216216216216216,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0),

OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335),

OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])


In [ ]:
set(data[(data['StockCode']=='22386') | (data['StockCode']=='85099B') | (data['StockCode']=='22699') | (data['StockCode']=='22697') ]['Description'])

{'GREEN REGENCY TEACUP AND SAUCER',
 'JUMBO BAG PINK POLKADOT',
 'JUMBO BAG RED RETROSPOT',
 'ROSES REGENCY TEACUP AND SAUCER '}

#HW 1 หา Frequent Item Set เเละ Association Rules
* ของช่วงเวลา Christmas & New Year (1 Dec - 31 Dec)
* ของประเทศในทวีป Europe
* ของช่วงเวลา ตรุษจีน ในทวีป Asia (15 jan - 15 Feb)
* ของช่วงเวลา Christmas & New Year ในทวีป Europe(1 Dec - 31 Dec)
* ของวันวาเลนไทน์ (14 Feb)
* จัดกลุ่มตามประเทศ+เดิอน

##1.ของช่วงเวลา Christmas & New Year (1 Dec - 31 Dec)

###เเยกวันเดือนปีออกจากเวลา

In [ ]:
data['InvoiceDate'][0].split()[0]

'12/01/2010'

In [ ]:
date = []
for i in range(len(data['InvoiceDate'])):
    x = data['InvoiceDate'][i].split()[0]
    date.append(x)

L = set(date)

In [ ]:
date[0].split('/')

['12', '01', '2010']

###เเยกวัน เดือน ปีออกจากกัน

In [ ]:
day =[]
month =[]
year =[]
for i in range(len(data['InvoiceDate'])):
  x = date[i].split('/')[0]
  y = date[i].split('/')[1]
  z = date[i].split('/')[2]
  month.append(x)
  day.append(y)
  year.append(z)

ซึ่งจะถือว่ารายการ 'วันที่' มีสตริงที่มีรูปแบบเช่น 'dd/mm/yyyy' (วัน/เดือน/ปี) ปรับโค้ดให้เหมาะสมหากรูปแบบวันที่ของคุณแตกต่างออกไป

หลังจากรันโค้ดนี้แล้ว คุณควรมีรายการ 'วัน', 'เดือน' และ 'ปี' แยกกัน ซึ่งมีค่าที่สอดคล้องกันซึ่งดึงมาจากรายการ 'วันที่'

###ใส่ดาต้าใหม่

In [ ]:
data_test_nonclane=data.copy()

In [ ]:
data_test_nonclane['date']=date
data_test_nonclane['day']=day
data_test_nonclane['month']=month
data_test_nonclane['year']=year

ดูเหมือนว่าโค้ดของคุณกำลังพยายามสร้าง DataFrame ใหม่ชื่อ data_test_nonclane ตามข้อมูล DataFrame ที่มีอยู่ จากนั้นเพิ่มคอลัมน์ใหม่ 'วันที่' 'วัน' 'เดือน' และ 'ปี' ลงไป

อย่างไรก็ตาม การมอบหมาย data_test_nonclane=data ทำให้ทั้งสองตัวแปรอ้างอิงออบเจ็กต์เดียวกัน ซึ่งหมายความว่าการเปลี่ยนแปลงใดๆ ในอันหนึ่งจะส่งผลต่ออีกอันหนึ่ง หากคุณต้องการสร้าง DataFrame ใหม่และเก็บข้อมูล DataFrame เดิมไว้ไม่เปลี่ยนแปลง คุณควรใช้เมธอด copy():

In [ ]:
#set(data_test_nonclane['day'])

###เเปลงให้ค่าเดือนอยู่ในหน่วยเดียวกัน

In [ ]:
data_test_nonclane[(data_test_nonclane['month']=='01')|(data_test_nonclane['month']=='02')|(data_test_nonclane['month']=='03')|(data_test_nonclane['month']=='04')|(data_test_nonclane['month']=='05')|(data_test_nonclane['month']=='06')|
                   (data_test_nonclane['month']=='07')|(data_test_nonclane['month']=='08')|(data_test_nonclane['month']=='09')]

InvoiceNo StockCode                         Description  Quantity  \
42481     539993     22386             JUMBO BAG PINK POLKADOT        10   
42482     539993     21499                  BLUE POLKADOT WRAP        25   
42483     539993     21498                 RED RETROSPOT WRAP         25   
42484     539993     22379            RECYCLING BAG RETROSPOT          5   
42485     539993     20718           RED RETROSPOT SHOPPER BAG        10   
...          ...       ...                                 ...       ...   
337783   C566471    85123A  WHITE HANGING HEART T-LIGHT HOLDER        -1   
337784   C566472     47566                       PARTY BUNTING        -8   
337785   C566473     82483  WOOD 2 DRAWER CABINET WHITE FINISH        -2   
337786   C566474     22960            JAM MAKING SET WITH JARS        -1   
337787   C566475         D                            Discount        -1   

             InvoiceDate  UnitPrice  CustomerID         Country        date  \
42481   01/04/2011 10:00       1.95     13313.0  United Kingdom  01/04/2011   
42482   01/04/2011 10:00       0.42     13313.0  United Kingdom  01/04/2011   
42483   01/04/2011 10:00       0.42     13313.0  United Kingdom  01/04/2011   
42484   01/04/2011 10:00       2.10     13313.0  United Kingdom  01/04/2011   
42485   01/04/2011 10:00       1.25     13313.0  United Kingdom  01/04/2011   
...                  ...        ...         ...             ...         ...   
337783  09/12/2011 18:18       2.95     16670.0  United Kingdom  09/12/2011   
337784  09/12/2011 18:19       4.95     16187.0  United Kingdom  09/12/2011   
337785  09/12/2011 18:25       6.95     16705.0  United Kingdom  09/12/2011   
337786  09/12/2011 18:26       4.25     15136.0  United Kingdom  09/12/2011   
337787  09/12/2011 18:26      20.75     14527.0  United Kingdom  09/12/2011   

       day month  year  
42481   04    01  2011  
42482   04    01  2011  
42483   04    01  2011  
42484   04    01  2011  
42485   04    01  2011  
...     ..   ...   ...  
337783  12    09  2011  
337784  12    09  2011  
337785  12    09  2011  
337786  12    09  2011  
337787  12    09  2011  

[127400 rows x 12 columns]

ดูเหมือนว่าคุณกำลังพยายามกรอง DataFrame data_test_nonclane ตามค่าในคอลัมน์ 'เดือน' อย่างไรก็ตาม ข้อมูลโค้ดที่ให้มานั้นไม่สมบูรณ์ และมีส่วนที่มีการใส่ความคิดเห็นไว้ หากคุณต้องการกรองแถวที่คอลัมน์ 'เดือน' มีค่าตั้งแต่ '01' ถึง '09' คุณสามารถลดความซับซ้อนของโค้ดโดยใช้เมธอด .isin() ได้ ดังนี้

In [ ]:
filtered_data = data_test_nonclane[data_test_nonclane['month'].isin(['01', '02', '03', '04', '05', '06', '07', '08', '09'])]

แปลงเดือน

In [ ]:
data_test=data_test_nonclane.replace(to_replace=["1", "2",'3','4','5','6','7','8','9','10','11','12','01','02','03','04','05','06','07','08','09'],value=[1, 2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9])

In [ ]:
data_test = data_test_nonclane.replace(to_replace=["1", "2", '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '01', '02', '03', '04', '05', '06', '07', '08', '09'],
                                       value=[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 1, 2, 3, 4, 5, 6, 7, 8, 9])


ดูเหมือนว่าคุณกำลังใช้วิธีแทนที่เพื่อแทนที่ค่าบางค่าใน DataFrame data_test_nonclane ความตั้งใจของคุณดูเหมือนจะแปลงค่าเดือนที่แสดงเป็นสตริงเป็นค่าตัวเลข

อย่างไรก็ตาม สิ่งสำคัญที่ควรทราบคือวิธีการแทนที่เพียงอย่างเดียวจะไม่แก้ไข DataFrame แบบแทนที่ เว้นแต่คุณจะตั้งค่าพารามิเตอร์ inplace=True หากคุณต้องการสร้าง DataFrame ใหม่ด้วยการแทนที่ คุณต้องกำหนดผลลัพธ์ให้กับตัวแปรใหม่

In [ ]:
set(data_test['day'])

{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12}

แปลงวัน

In [ ]:
x=[]
for i in range(1,32):
  x.append(i)
print(x)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31]


In [ ]:
data_test1=data_test_nonclane.replace(['01', '02', '03', '04', '05', '06', '07', '08', '09', '10','11',
                                      '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '30', '31'],x)

In [ ]:
set(data_test1['month'])

{1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31}

In [ ]:
data_test=data_test1.replace(to_replace=["1", "2",'3','4','5','6','7','8','9'],value=[1,2,3,4,5,6,7,8,9])

In [ ]:
data_test.to_csv('how.csv')

In [ ]:
x = data_test[(data_test['day'] == 14) & (data_test['month'] == 2)]
print(x)

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country, date, day, month, year]
Index: []


###เริ่มทำข้อ 1


In [ ]:
data_C_and_N = data_test[data_test['month']==12]

In [ ]:
#!pip install apyori
#from apyori import apriori

In [ ]:
transactionsC_and_N = data_C_and_N.groupby('InvoiceNo')['StockCode'].apply(list)

In [ ]:
frequentItemsets = list(apriori(transactionsC_and_N,min_support = 0.05))
#frequentItemsets

In [ ]:
min_sup = 0.04

frequentItemsetsC_and_N = list(apriori(transactionsC_and_N,min_support = min_sup))
for i in frequentItemsetsC_and_N:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22086', '22910'}), support=0.047697368421052634, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22086', '22910'}), confidence=0.047697368421052634, lift=1.0), OrderedStatistic(items_base=frozenset({'22086'}), items_add=frozenset({'22910'}), confidence=0.5160142348754448, lift=7.195794834960331), OrderedStatistic(items_base=frozenset({'22910'}), items_add=frozenset({'22086'}), confidence=0.6651376146788991, lift=7.195794834960332)])


In [ ]:
transactionsC_and_N = data_C_and_N.groupby(['Country', 'month'])['StockCode'].apply(list)

In [ ]:
min_sup =0.3
frequentItemsetsC_and_N = list(apriori(transactionsC_and_N,min_support = min_sup))
for i in frequentItemsetsC_and_N:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'POST', '20750'}), support=0.3333333333333333, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'POST', '20750'}), confidence=0.3333333333333333, lift=1.0), OrderedStatistic(items_base=frozenset({'20750'}), items_add=frozenset({'POST'}), confidence=0.9, lift=1.7357142857142858), OrderedStatistic(items_base=frozenset({'POST'}), items_add=frozenset({'20750'}), confidence=0.6428571428571429, lift=1.735714285714286)])
yes
RelationRecord(items=frozenset({'22728', '22727'}), support=0.3333333333333333, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22728', '22727'}), confidence=0.3333333333333333, lift=1.0), OrderedStatistic(items_base=frozenset({'22727'}), items_add=frozenset({'22728'}), confidence=1.0, lift=3.0), OrderedStatistic(items_base=frozenset({'22728'}), items_add=frozenset({'22727'}), confidence=1.0, lift=3.0)])


สรุปได้ดังนั้น
* RelationRecord(items=frozenset({'22910', '22086'}), support=0.047697368421052634)
  * ordered_statistics=OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22910', '22086'}), confidence=0.047697368421052634, lift=1.0)
  * OrderedStatistic(items_base=frozenset({'22086'}), items_add=frozenset({'22910'}), confidence=0.5160142348754448, lift=7.195794834960331)
  * OrderedStatistic(items_base=frozenset({'22910'}), items_add=frozenset({'22086'}), confidence=0.6651376146788991, lift=7.195794834960332)


In [ ]:
def finditime(x,y) :
  L=set(data[(data['StockCode']==x)|(data['StockCode']==y)]['Description'])
  return L

In [ ]:
finditime('22910','22086')

{"PAPER CHAIN KIT 50'S CHRISTMAS ", 'PAPER CHAIN KIT VINTAGE CHRISTMAS'}

##2.ของประเทศในทวีป Europe

###เเบ่งทวีป

In [ ]:
set(data['Country'])

{'Australia',
 'Austria',
 'Bahrain',
 'Belgium',
 'Brazil',
 'Canada',
 'Channel Islands',
 'Cyprus',
 'Czech Republic',
 'Denmark',
 'EIRE',
 'European Community',
 'Finland',
 'France',
 'Germany',
 'Greece',
 'Hong Kong',
 'Iceland',
 'Israel',
 'Italy',
 'Japan',
 'Lebanon',
 'Lithuania',
 'Malta',
 'Netherlands',
 'Norway',
 'Poland',
 'Portugal',
 'RSA',
 'Saudi Arabia',
 'Singapore',
 'Spain',
 'Sweden',
 'Switzerland',
 'USA',
 'United Arab Emirates',
 'United Kingdom',
 'Unspecified'}

ทวีปยุโรป

In [ ]:
UP_country = ['Austria','Belgium','Cyprus','Czech Republic','Denmark','EIRE','European Community',
              'Finland','France','Germany','Greece','Iceland','Italy','Lithuania','Malta','Netherlands',
              'Norway','Poland','Portugal','Spain','Sweden','United Kingdom']

ทวีปเอเชีย

In [ ]:
Asia_country = ['Hong Kong','Bahrain','Israel','Japan','Lebanon','Saudi Arabia','Singapore','United Arab Emirates']

In [ ]:
#a0=data_test[data_test['Country']==UP_country[0]]

### หาข้อมูลทวีปยุโรป

In [ ]:
EU_data = data_test[data_test['Country']==""]
for i in range(len(UP_country)):
  x = data_test[data_test['Country']==UP_country[i]]
  EU_data=pd.concat([x,EU_data])

In [ ]:
len(set(EU_data['Country']))

22

###หาข้อ 2

In [ ]:
transactionsEU = EU_data.groupby('InvoiceNo')['StockCode'].apply(list)

In [ ]:
frequentItemsets = list(apriori(transactionsEU,min_support = 0.05))
#frequentItemsets

In [ ]:
min_sup = 0.03

frequentItemsetsEU= list(apriori(transactionsEU,min_support = min_sup))
for i in frequentItemsetsEU:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22386', '85099B'}), support=0.03243053387449266, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03243053387449266, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6772616136919316, lift=8.201394890946151), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39272211720226846, lift=8.201394890946151)])
yes
RelationRecord(items=frozenset({'22699', '22697'}), support=0.030049953168904152, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.030049953168904152, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7389635316698656, lift=17.213819577735123), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.21381957773

In [ ]:
finditime('22386','85099B')

{'JUMBO BAG PINK POLKADOT', 'JUMBO BAG RED RETROSPOT'}

##3.ของช่วงเวลา ตรุษจีน ในทวีป Asia (15 jan - 15 Feb)

###เเยกทวีปเอเชีย

In [ ]:
Asia_data = data_test[data_test['Country']==""]
for i in range(len(Asia_country)):
  x = data_test[data_test['Country']==Asia_country[i]]
  Asia_data=pd.concat([x,Asia_data])

In [ ]:
len(set(Asia_data['Country']))

8

###หาข้อมูลช่วงตรุษจีน

In [ ]:
M1_Asia_data=Asia_data[(Asia_data['month']==1)]

In [ ]:
D1_Asia_data=M1_Asia_data[(M1_Asia_data['day']>=15)]

In [ ]:
#D1_Asia_data

In [ ]:
M2_Asia_data=Asia_data[(Asia_data['month']==2)]

In [ ]:
#M2_Asia_data

In [ ]:
D2_Asia_data=M2_Asia_data[(M2_Asia_data['day']<=15)]

In [ ]:
D2_Asia_data

InvoiceNo StockCode                          Description  Quantity  \
89570    543911     21485     RETROSPOT HEART HOT WATER BOTTLE         6   
89571    543911     48138                   DOORMAT UNION FLAG         2   
89572    543911     85150        LADIES & GENTLEMEN METAL SIGN        24   
89573    543911     82552                  WASHROOM METAL SIGN        24   
89574    543911     82551               LAUNDRY 15C METAL SIGN        12   
...         ...       ...                                  ...       ...   
85761    543518     21080  SET/20 RED RETROSPOT PAPER NAPKINS         96   
85762    543518     21217        RED RETROSPOT ROUND CAKE TINS        24   
85763    543518     21216     SET 3 RETROSPOT TEA,COFFEE,SUGAR        48   
85764    543518     21980    PACK OF 12 RED RETROSPOT TISSUES         96   
85765    543518     22333    RETROSPOT PARTY BAG + STICKER SET        96   

            InvoiceDate  UnitPrice  CustomerID               Country  \
89570   2/14/2011 12:46       4.95     17829.0  United Arab Emirates   
89571   2/14/2011 12:46       7.95     17829.0  United Arab Emirates   
89572   2/14/2011 12:46       2.55     17829.0  United Arab Emirates   
89573   2/14/2011 12:46       1.45     17829.0  United Arab Emirates   
89574   2/14/2011 12:46       1.45     17829.0  United Arab Emirates   
...                 ...        ...         ...                   ...   
85761  02/09/2011 11:21       0.64     12753.0                 Japan   
85762  02/09/2011 11:21       8.95     12753.0                 Japan   
85763  02/09/2011 11:21       4.25     12753.0                 Japan   
85764  02/09/2011 11:21       0.29     12753.0                 Japan   
85765  02/09/2011 11:21       1.45     12753.0                 Japan   

             date  day  month  year  
89570   2/14/2011   14      2  2011  
89571   2/14/2011   14      2  2011  
89572   2/14/2011   14      2  2011  
89573   2/14/2011   14      2  2011  
89574   2/14/2011   14      2  2011  
...           ...  ...    ...   ...  
85761  02/09/2011    9      2  2011  
85762  02/09/2011    9      2  2011  
85763  02/09/2011    9      2  2011  
85764  02/09/2011    9      2  2011  
85765  02/09/2011    9      2  2011  

[115 rows x 12 columns]

###เอาสองตารางรวมกัน

In [ ]:
use_Asia_data=pd.concat([D1_Asia_data,D2_Asia_data])

In [ ]:
use_Asia_data

InvoiceNo StockCode                          Description  Quantity  \
70758    542102     21519      GIN & TONIC DIET GREETING CARD         72   
70759    542102     22697      GREEN REGENCY TEACUP AND SAUCER         6   
70760    542102     22699     ROSES REGENCY TEACUP AND SAUCER          6   
70761    542102     22343       PARTY PIZZA DISH RED RETROSPOT        24   
70762    542102     22344       PARTY PIZZA DISH PINK POLKADOT        24   
...         ...       ...                                  ...       ...   
85761    543518     21080  SET/20 RED RETROSPOT PAPER NAPKINS         96   
85762    543518     21217        RED RETROSPOT ROUND CAKE TINS        24   
85763    543518     21216     SET 3 RETROSPOT TEA,COFFEE,SUGAR        48   
85764    543518     21980    PACK OF 12 RED RETROSPOT TISSUES         96   
85765    543518     22333    RETROSPOT PARTY BAG + STICKER SET        96   

            InvoiceDate  UnitPrice  CustomerID    Country        date  day  \
70758   1/25/2011 13:26       0.36     12744.0  Singapore   1/25/2011   25   
70759   1/25/2011 13:26       2.95     12744.0  Singapore   1/25/2011   25   
70760   1/25/2011 13:26       2.95     12744.0  Singapore   1/25/2011   25   
70761   1/25/2011 13:26       0.21     12744.0  Singapore   1/25/2011   25   
70762   1/25/2011 13:26       0.21     12744.0  Singapore   1/25/2011   25   
...                 ...        ...         ...        ...         ...  ...   
85761  02/09/2011 11:21       0.64     12753.0      Japan  02/09/2011    9   
85762  02/09/2011 11:21       8.95     12753.0      Japan  02/09/2011    9   
85763  02/09/2011 11:21       4.25     12753.0      Japan  02/09/2011    9   
85764  02/09/2011 11:21       0.29     12753.0      Japan  02/09/2011    9   
85765  02/09/2011 11:21       1.45     12753.0      Japan  02/09/2011    9   

       month  year  
70758      1  2011  
70759      1  2011  
70760      1  2011  
70761      1  2011  
70762      1  2011  
...      ...   ...  
85761      2  2011  
85762      2  2011  
85763      2  2011  
85764      2  2011  
85765      2  2011  

[273 rows x 12 columns]

###หาข้อ 3

In [ ]:
transactionsAsia = use_Asia_data.groupby('InvoiceNo')['StockCode'].apply(list)

In [ ]:
transactionsAsia

InvoiceNo
541975    [22620, 85183B, 85183A, 16236, 21058, 21967, 2...
542102    [21519, 22697, 22699, 22343, 22344, 22345, 223...
542276    [82551, 21165, 82600, 21754, 21756, 82482, 824...
543179    [22407, 22406, 22405, 21745, 22496, 22495, 224...
543518    [22759, 22662, 22661, 21717, 22957, 22861, 222...
543911    [21485, 48138, 85150, 82552, 82551, 82583, 224...
Name: StockCode, dtype: object

In [ ]:
min_sup = 0.32
frequentItemsets_A= list(apriori(transactionsAsia,min_support = min_sup))
for i in frequentItemsets_A:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22202', '20682'}), support=0.3333333333333333, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20682', '22202'}), confidence=0.3333333333333333, lift=1.0), OrderedStatistic(items_base=frozenset({'20682'}), items_add=frozenset({'22202'}), confidence=1.0, lift=3.0), OrderedStatistic(items_base=frozenset({'22202'}), items_add=frozenset({'20682'}), confidence=1.0, lift=3.0)])
yes
RelationRecord(items=frozenset({'22203', '20682'}), support=0.3333333333333333, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20682', '22203'}), confidence=0.3333333333333333, lift=1.0), OrderedStatistic(items_base=frozenset({'20682'}), items_add=frozenset({'22203'}), confidence=1.0, lift=3.0), OrderedStatistic(items_base=frozenset({'22203'}), items_add=frozenset({'20682'}), confidence=1.0, lift=3.0)])
yes
RelationRecord(items=frozenset({'20682', '22551'}), support=0.3333333333333333, ordered_statistics=[Or

##4.ของช่วงเวลา Christmas & New Year ในทวีป Europe(1 Dec - 31 Dec)

In [ ]:
C_N_EU_data=EU_data[EU_data['month']==12]

In [ ]:
transactionsC_N_EU = C_N_EU_data.groupby('InvoiceNo')['StockCode'].apply(list)

In [ ]:
transactionsC_N_EU

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 3022, dtype: object

In [ ]:
min_sup = 0.04
frequentItemsetsC_N_EU= list(apriori(transactionsC_N_EU,min_support = min_sup))
for i in frequentItemsetsC_N_EU:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22086', '22910'}), support=0.04765056254136334, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22086', '22910'}), confidence=0.04765056254136334, lift=1.0), OrderedStatistic(items_base=frozenset({'22086'}), items_add=frozenset({'22910'}), confidence=0.5142857142857143, lift=7.162080315997367), OrderedStatistic(items_base=frozenset({'22910'}), items_add=frozenset({'22086'}), confidence=0.663594470046083, lift=7.162080315997367)])


In [ ]:
finditime('22086', '22910')

{"PAPER CHAIN KIT 50'S CHRISTMAS ", 'PAPER CHAIN KIT VINTAGE CHRISTMAS'}

##5.ของวันวาเลนไทน์ (14 Feb)

In [ ]:
M14_data = data_test[(data_test['month']==2)]

In [ ]:
va_data=M14_data[M14_data['day']==14]

In [ ]:
transactions_va = va_data.groupby('InvoiceNo')['StockCode'].apply(list)

In [ ]:
min_sup = 0.09
frequentItemsets_va= list(apriori(transactions_va,min_support = min_sup))
for i in frequentItemsets_va:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22722', '22720'}), support=0.09090909090909091, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22722', '22720'}), confidence=0.09090909090909091, lift=1.0), OrderedStatistic(items_base=frozenset({'22720'}), items_add=frozenset({'22722'}), confidence=0.625, lift=5.729166666666667), OrderedStatistic(items_base=frozenset({'22722'}), items_add=frozenset({'22720'}), confidence=0.8333333333333334, lift=5.729166666666667)])
yes
RelationRecord(items=frozenset({'22969', '22720'}), support=0.09090909090909091, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22720', '22969'}), confidence=0.09090909090909091, lift=1.0), OrderedStatistic(items_base=frozenset({'22720'}), items_add=frozenset({'22969'}), confidence=0.625, lift=6.875), OrderedStatistic(items_base=frozenset({'22969'}), items_add=frozenset({'22720'}), confidence=1.0, lift=6.875)])


##6.จัดกลุ่มตามประเทศ+เดิอน

In [ ]:
transactions_Mon_Con = data_test.groupby(['Country', 'month'])['StockCode'].apply(list)

In [ ]:
transactions_Mon_Con

Country      month
Australia    1        [22078, 22079, 22080, 22077, 22505, 22516, 225...
             2        [84978, 22960, 47590B, 47566, 22087, 22720, 22...
             3        [22978, 22980, 22993, 22654, 22968, 22505, 225...
             4        [POST, 22232, 22993, 22449, 20726, 20725, 2072...
             5        [22699, 22859, 22730, 84663A, 23206, 48129, 23...
                                            ...                        
Unspecified  6        [23122, 23119, 23108, 23240, 23237, 23239, 232...
             7        [21634, 21212, 22964, 23254, 23256, 84378, 209...
             8        [23236, 21210, 23313, 23318, 22952, 22950, 232...
             9        [22904, 21329, 21992, 20772, 23196, 22543, 225...
             11       [22326, 20725, 22628, 22631, 84997D, 23254, 23...
Name: StockCode, Length: 296, dtype: object

In [ ]:
min_sup = 0.29
frequentItemsets_Mon_Con= list(apriori(transactions_Mon_Con,min_support = min_sup))
for i in frequentItemsets_Mon_Con:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22423', '21212'}), support=0.2905405405405405, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423', '21212'}), confidence=0.2905405405405405, lift=1.0), OrderedStatistic(items_base=frozenset({'21212'}), items_add=frozenset({'22423'}), confidence=0.7818181818181817, lift=1.7016042780748661), OrderedStatistic(items_base=frozenset({'22423'}), items_add=frozenset({'21212'}), confidence=0.6323529411764705, lift=1.7016042780748661)])
yes
RelationRecord(items=frozenset({'POST', '22326'}), support=0.32094594594594594, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'POST', '22326'}), confidence=0.32094594594594594, lift=1.0), OrderedStatistic(items_base=frozenset({'22326'}), items_add=frozenset({'POST'}), confidence=0.8407079646017699, lift=1.4638209266007287), OrderedStatistic(items_base=frozenset({'POST'}), items_add=frozenset({'22326'}), confidence=0.5588235294117647, lift=1.46382092

#quiz 22386 เเละ 85099B คือสินค้าอะไร

In [ ]:
set(data[data['StockCode']=='22386']['Description'])

{'JUMBO BAG PINK POLKADOT'}

In [ ]:
set(data[data['StockCode']=='85099B']['Description'])

{'JUMBO BAG RED RETROSPOT'}

RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03027027027027027, lift=1.0),

OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947),

OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])

In [ ]:
set(data[(data['StockCode']=='22699') | (data['StockCode']=='22697')]['InvoiceNo'])

{'549305',
 '570459',
 '550458',
 '567631',
 '547846',
 '566607',
 '556434',
 '562770',
 '573511',
 '575206',
 '580610',
 '544660',
 '548559',
 '554030',
 '578347',
 '562102',
 '563166',
 '563910',
 '543399',
 '548768',
 '548739',
 '546562',
 '550989',
 '544456',
 '554113',
 '554039',
 '578065',
 '546660',
 '563384',
 '541506',
 '548734',
 '545759',
 '562607',
 '559493',
 '568048',
 '546837',
 '536523',
 '553462',
 'C545438',
 '545078',
 '551995',
 '572854',
 '545214',
 '562950',
 '562374',
 '556812',
 '562933',
 '562553',
 'C546865',
 '557788',
 '539492',
 '563909',
 '541969',
 '551997',
 '563169',
 '543055',
 '540992',
 '546367',
 '581433',
 '568065',
 '558443',
 '554337',
 '573020',
 '559295',
 '549525',
 '559827',
 '574707',
 '541866',
 '570677',
 '549852',
 '542619',
 '561702',
 '539564',
 '557642',
 '548260',
 '580983',
 '568134',
 'C540393',
 'C566084',
 '547712',
 '568955',
 '543201',
 '545429',
 '565234',
 '575185',
 '576899',
 '544339',
 '561645',
 '576616',
 '539593',
 '5627

In [ ]:
data[data['InvoiceNo']=='562950']

InvoiceNo StockCode                         Description  Quantity  \
297488    562950     22752        SET 7 BABUSHKA NESTING BOXES         2   
297489    562950     22755     SMALL PURPLE BABUSHKA NOTEBOOK         12   
297490    562950     22753     SMALL YELLOW BABUSHKA NOTEBOOK         12   
297491    562950     22754        SMALL RED BABUSHKA NOTEBOOK         12   
297492    562950     22757        LARGE RED BABUSHKA NOTEBOOK         12   
297493    562950     22758    LARGE PURPLE BABUSHKA NOTEBOOK          12   
297494    562950     22756     LARGE YELLOW BABUSHKA NOTEBOOK         12   
297495    562950     23310             BUBBLEGUM RING ASSORTED        36   
297496    562950     23280  FOLDING BUTTERFLY MIRROR HOT PINK         12   
297497    562950     22383              LUNCH BAG SUKI DESIGN         10   
297498    562950     22492             MINI PAINT SET VINTAGE         36   
297499    562950     71459      HANGING JAM JAR T-LIGHT HOLDER        12   
297500    562950     84945  MULTI COLOUR SILVER T-LIGHT HOLDER        12   
297501    562950     21175         GIN + TONIC DIET METAL SIGN        12   
297502    562950     23164                  REGENCY CAKE SLICE         4   
297503    562950     23245          SET OF 3 REGENCY CAKE TINS         4   
297504    562950     23171            REGENCY TEA PLATE GREEN         12   
297505    562950     22697     GREEN REGENCY TEACUP AND SAUCER         6   

             InvoiceDate  UnitPrice  CustomerID         Country        date  \
297488  08/11/2011 09:56       8.50     15608.0  United Kingdom  08/11/2011   
297489  08/11/2011 09:56       0.85     15608.0  United Kingdom  08/11/2011   
297490  08/11/2011 09:56       0.85     15608.0  United Kingdom  08/11/2011   
297491  08/11/2011 09:56       0.85     15608.0  United Kingdom  08/11/2011   
297492  08/11/2011 09:56       1.25     15608.0  United Kingdom  08/11/2011   
297493  08/11/2011 09:56       1.25     15608.0  United Kingdom  08/11/2011   
297494  08/11/2011 09:56       1.25     15608.0  United Kingdom  08/11/2011   
297495  08/11/2011 09:56       0.42     15608.0  United Kingdom  08/11/2011   
297496  08/11/2011 09:56       0.83     15608.0  United Kingdom  08/11/2011   
297497  08/11/2011 09:56       1.65     15608.0  United Kingdom  08/11/2011   
297498  08/11/2011 09:56       0.65     15608.0  United Kingdom  08/11/2011   
297499  08/11/2011 09:56       0.85     15608.0  United Kingdom  08/11/2011   
297500  08/11/2011 09:56       0.85     15608.0  United Kingdom  08/11/2011   
297501  08/11/2011 09:56       2.55     15608.0  United Kingdom  08/11/2011   
297502  08/11/2011 09:56       4.95     15608.0  United Kingdom  08/11/2011   
297503  08/11/2011 09:56       4.95     15608.0  United Kingdom  08/11/2011   
297504  08/11/2011 09:56       1.65     15608.0  United Kingdom  08/11/2011   
297505  08/11/2011 09:56       2.95     15608.0  United Kingdom  08/11/2011   

       day month  year  
297488  11    08  2011  
297489  11    08  2011  
297490  11    08  2011  
297491  11    08  2011  
297492  11    08  2011  
297493  11    08  2011  
297494  11    08  2011  
297495  11    08  2011  
297496  11    08  2011  
297497  11    08  2011  
297498  11    08  2011  
297499  11    08  2011  
297500  11    08  2011  
297501  11    08  2011  
297502  11    08  2011  
297503  11    08  2011  
297504  11    08  2011  
297505  11    08  2011